libyul/AST.h

```diff
@@ -35,49 +35,67 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct TypedName { langutil::SourceLocation location; YulString name; Type type; };
+struct DebugData
+{
+	explicit DebugData(langutil::SourceLocation _location) : location(std::move(_location)) {}
+	langutil::SourceLocation location;
+};
+
+inline std::shared_ptr<DebugData const> createDebugData(langutil::SourceLocation _location = {})
+{
+	return std::make_shared<DebugData const>(_location);
+}
+
+inline void changeDebugDataLocationEnd(std::shared_ptr<DebugData const>& io_debugData, int _end)
+{
+	langutil::SourceLocation loc = io_debugData->location;
+	loc.end = _end;
+	io_debugData = std::make_shared<DebugData const>(loc);
+}
+
+struct TypedName { std::shared_ptr<DebugData const> debugData; YulString name; Type type; };
 using TypedNameList = std::vector<TypedName>;
 
 /// Literal number or string (up to 32 bytes)
 enum class LiteralKind { Number, Boolean, String };
-struct Literal { langutil::SourceLocation location; LiteralKind kind; YulString value; Type type; };
+struct Literal { std::shared_ptr<DebugData const> debugData; LiteralKind kind; YulString value; Type type; };
```

Comment:
Why did we decide to use shared_ptr and not just having the debugdata as a value here? Do we think we'll be able to share these debugdata pointers between different items?

Reply:
Yes, I think especially when modifying the AST during optimization, this is useful because the debug data does not have to be copied.

Reply:
I think it should stay as `shared_ptr` as it is going to be re-used by optimizer at least